fix(extui): close cmdwin to enter "more" window

luukvbaal opened the pull request:
Problem:  Cannot enter "more" window while cmdwin is open since it is a
          regular window and changing windows is disallowed.
Solution: Close cmdwin to enter the "more" window. Minor regression
          w.r.t. the current message grid. Resolving that will require
          somehow bypassing textlock for the "more" window.

Fix #33853

luukvbaal commented:
🫣

justinmk commented:
wdyt @seandewar  ? 

is this RFC and then tests/docs will be updated?

seandewar commented on src/nvim/ex_docmd.c:
```diff
@@ -4732,7 +4732,7 @@ bool before_quit_autocmds(win_T *wp, bool quit_all, bool forceit)
 static void ex_quit(exarg_T *eap)
 {
-  if (cmdwin_type != 0) {
+  if (curwin == cmdwin_win) {
```
Can use this to close other windows when in the messages window, which can allow for `cmdwin_old_curwin` to be closed or the cmdwin to be left as the only window; that can cause strangeness.

E.g: this causes E199, which shouldn't happen:

```vim
only!
lua require"vim._extui".enable{}
echom "foo"
call feedkeys("q:", "nt")
au CmdwinEnter * ++once mess | 1q! | qa! | au SafeState * ++once au SafeState * ++once mes
```

seandewar commented on src/nvim/ex_docmd.c:
```diff
@@ -4845,7 +4845,7 @@ static void ex_close(exarg_T *eap)
   win_T *win = NULL;
   int winnr = 0;
-  if (cmdwin_type != 0) {
+  if (curwin == cmdline_win) {
```
Same here:

```vim
only!
lua require"vim._extui".enable{}
echom "foo"
call feedkeys("q:", "nt")
au CmdwinEnter * ++once mess | 1close! | qa! | au SafeState * ++once au SafeState * ++once mes
```

seandewar commented on src/nvim/globals.h:
```diff
@@ -743,6 +743,7 @@ EXTERN bool km_startsel INIT( = false);
 EXTERN int cmdwin_type INIT( = 0);    ///< type of cmdline window or 0
 EXTERN int cmdwin_result INIT( = 0);  ///< result of cmdline window or 0
 EXTERN int cmdwin_level INIT( = 0);   ///< cmdline recursion level
+EXTERN bool cmdwin_lock INIT( = true);   ///< should lock for open cmdwin
```
With my suggestion from earlier, might be better to make this `cmdwin_nolock_win`/`buf` instead (can bikeshed the naming), and only allow navigation between that window/buffer and the cmdwin?

seandewar commented on src/nvim/window.c:
```diff
@@ -213,7 +213,7 @@ void do_window(int nchar, int Prenum, int xchar)
 #define CHECK_CMDWIN \
   do { \
-    if (cmdwin_type != 0) { \
+    if (curwin == cmdwin_win && cmdwin_lock) { \
```
Likewise, this allows entering other windows once in the messages window with `<C-W>w` and the like.

seandewar reviewed (commented):
Haven't had time to have a thorough look yet, but here's a few observations.

I commend your bravery, soldier :saluting_face:

luukvbaal commented on src/nvim/ex_getln.c:
```diff
@@ -2930,7 +2930,7 @@ int check_opt_wim(void)
 bool text_locked(void)
 {
-  if (cmdwin_type != 0) {
+  if (curwin == cmdwin_win && cmdwin_lock) {
```
Thanks for the review! I remember contemplating 'winfixbuf' but don't remember the reason for omitting it, if I had one😓 The correct buffer is restored for all extui windows on the next event anyways so 'switching the buffer shouldn't break anything but I also cant think of a reason to allow it.

luukvbaal committed:
fix(extui): close cmdwin to enter "more" window

Problem:  Cannot enter "more" window while cmdwin is open since it is a
          regular window and changing windows is disallowed.
Solution: Close cmdwin to enter the "more" window. Minor regression
          w.r.t. the current message grid. Resolving that will require
          somehow bypassing textlock for the "more" window.

seandewar commented on runtime/lua/vim/_extui/messages.lua:
```diff
@@ -405,17 +405,24 @@ function M.set_pos(type)
-      api.nvim_set_current_win(win)
+      -- Cannot leave the cmdwin to enter the "more" window, so close it.
+      if fn.getcmdwintype() ~= '' then
+        api.nvim_win_close(0, true)
```
A _slightly_ pedantic nitpick (sorry! :grin:), but it's possible for the cmdwin to not be curwin (but only temporarily).

E.g: (run this one manually)

```vim
lua require'vim._extui'.enable{}
let s:w = win_getid()
call feedkeys('q:', 'nt')
au CmdwinEnter * ++once call win_execute(s:w, 'messages')
```

_Should_ be able to get around that by just using a plain `:quit` instead (it prioritizes the cmdwin regardless of curwin), but looping through the list of windows to `nvim_win_close` the one with `win_gettype() == "command"` may be more robust.

Yeah, the cmdwin is a bit jank...
